compiler/rustc_infer/src/infer/canonical/canonicalizer.rs

```diff
@@ -487,7 +502,19 @@ impl<'cx, 'tcx> TypeFolder<TyCtxt<'tcx>> for Canonicalizer<'cx, 'tcx> {
                 }
             }
             ty::ConstKind::Infer(InferConst::EffectVar(vid)) => {
-                match self.infcx.unwrap().probe_effect_var(vid) {
+                let Some(infcx) = self.infcx else {
```

Comment:
I've avoided to review this PR for a while as this feels very suspicious to me but I wasn't able to explain why without a significant time investment. sorry 😅 

The ICE occurs in `equate_impl_headers` which is used to compute the constraints we can assume. We cannot eagerly replace infer vars with placeholders at this point https://github.com/rust-lang/rust/blob/d86e1229411c086e1267c80dd9872959ca13b8b9/compiler/rustc_trait_selection/src/traits/coherence.rs#L449

We should use an `infcx` when encountering inference varialbes to make sure these infer vars do not resolve to any other types. I would personally prefer to only do the `deeply_normalize` change in this PR and fix the `negative_coherence` issue by cleaning up our const handling in general.

While your approach does work and may at worst slightly worsen type inference here, this change does cause the code to be "conceptionally broken"[^1], so I would prefer to not land it, esp given that it's an ICE which only occurs with a nightly feature enabled.

[^1]: this is mostly just vibe based 🤔 using it here as the way to rationalize this change is "yes, this may cause unexpected and wrong results in rare cases, but it's not unsound, so it's fine". It's somewhat difficult to explain exactly what I mean here

Reply:
For context, I tried to add a subdiagnostic “if there's a conflicting implementation and if using negative coherence solves it, suggests adding `#![feature(with_negative_coherence)]`” but it ICEd some tests, and I may have been too hasty in my attempt to fix it. I’ll try to come up with a nicer approach when I have time.